byzcoin/viewchange.go

```diff
@@ -330,26 +330,25 @@ func (s *Service) createViewChangeBlock(req viewchange.NewViewReq, multisig []by
 		return errors.New("roster size is too small, must be >= 4")
 	}
 
-	st, err := s.GetReadOnlyStateTrie(req.GetGen())
+	reqBuf, err := protobuf.Encode(&req)
 	if err != nil {
 		return err
 	}
-	_, _, genDarcID, err := st.GetValues(NewInstanceID(nil).Slice())
+
+	signer := darc.NewSignerEd25519(s.ServerIdentity().Public, s.getPrivateKey())
+
+	st, err := s.GetReadOnlyStateTrie(sb.SkipChainID())
 	if err != nil {
 		return err
 	}
-
-	reqBuf, err := protobuf.Encode(&req)
+	ctr, err := getSignerCounter(st, signer.Identity().String())
```

Comment:
Thought experiment: leader fails, second in roster uses this code to look up their own last counter, and signs the view change. Second fails. Third notices heartbeat failure, does the same thing, looking up their own counter. Fourth, fifth, etc have seen second's transaction, and signed it, but quorum was not reached. So updateTrie was never called, so second's counter is not updated. Third's view change goes through. Much later (once it is second's turn again to become leader), second needs to create a view change again, and calls getSignerCounter again. Because no txn with second's counter++ was ever committed into the blockchain, second will get the same counter again. But everyone will be ok with that.

So this is a case of counter reuse, but correct and desired?